File: python/python/ch7/PerformanceChecker.py
import pydash

from time import time
from enum import Enum

class CheckerTimeType(Enum):
    sec = 'sec'
    ms = 'ms'
    us = 'us'
        
class PerformanceChecker:
    def __init__(self, name: str, accumulate_mode: bool, digit: int):
        self.begin = 0
        self.end = 0
        self.name = name
        self.accumulate_mode = accumulate_mode
        self.accumulate_time = 0
        self.digit = digit
      
    @property
    def difference(self):
        return self.end - self.begin
    
def reset_checker_accumulate(checker: PerformanceChecker):
    return PerformanceChecker(checker.name, False, checker.digit)

def start_checker_curry(accumulate_mode: bool, digit: int, name: str):
    checker = PerformanceChecker(name, accumulate_mode, digit)
    checker.begin = time()
    return checker

start_checker = pydash.curry(start_checker_curry)

def stop_checker(checker: PerformanceChecker):
    checker.end = time()
    return checker

def diff_checker(checker: PerformanceChecker):
    return checker.difference if checker.accumulate_mode == False else checker.accumulate_time

def diff_string_checker_curry(checker_time_type: CheckerTimeType, checker: PerformanceChecker):
    diff = diff_checker(checker)
    unit = 's'
    if checker_time_type == CheckerTimeType.ms:
        diff *= 1000
        unit = 'ms'
    elif checker_time_type == CheckerTimeType.us:
        diff *= 1000000
        unit = 'us'
    return f'{diff:.{checker.digit}f}{unit}'

diff_string_checker = pydash.curry(diff_string_checker_curry)

def log_checker_curry(checker_time_type: CheckerTimeType, checker: PerformanceChecker):
    print(f'{checker.name}: {diff_string_checker(checker_time_type, checker)}')
    
log_checker = pydash.curry(log_checker_curry)

def map_between_checker_curry(func, checker: PerformanceChecker):
    func()
    return checker

map_between_checker = pydash.curry(map_between_checker_curry)

File: javascript/ch7/PerformanceChecker.js
const fp = require('lodash/fp');

class CheckerTimeType {
  static sec = 'sec';
  static ms = 'ms';
  static us = 'us';
}

class PerformanceChecker {
  constructor(name, accumulateMode, digit) {
    this.begin = 0;
    this.end = 0;
    this.name = name;
    this.accumulateMode = accumulateMode;
    this.accumulateTime = 0;
    this.digit = digit;
  }

  get difference() {
    return this.end - this.begin;
  }
}

const propSet = fp.curry((obj, prop, data) => {
  obj[prop] = data;
  return obj;
});

const currentTime = () => Date.now();

const resetCheckerAccumulate = (checker) => new PerformanceChecker(checker.name, false, checker.digit);

const startChecker = fp.curry((accumulateMode, digit, name) => 
  propSet(new PerformanceChecker(name, accumulateMode, digit), 'begin', currentTime()));

const stopChecker = (checker) => 
  propSet(checker, 'end', currentTime());

const diffChecker = (checker) => 
  checker.accumulateMode == false ? checker.difference : checker.accumulateTime;

const diffStringChecker = fp.curry((checkerTimeType, checker) => {
  let diff = diffChecker(checker);
  let unit = 'ms';
  switch (checkerTimeType) {
    case CheckerTimeType.sec:
      diff /= 1000;
      unit = 's';
      break;
    case CheckerTimeType.us:
      diff *= 1000;
      unit = 'us';
      break;
  }
  return `${diff.toFixed(checker.digit)}${unit}`;
});

const logChecker = fp.curry((checkerTimeType, checker) => 
  console.log(`${checker.name}: ${diffStringChecker(checkerTimeType, checker)}`));

const mapBetweenChecker = fp.curry((func, checker) => {
  func();
  return checker;
});

module.exports = {
  CheckerTimeType,
  PerformanceChecker,
  resetCheckerAccumulate,
  startChecker,
  stopChecker,
  diffChecker,
  diffStringChecker,
  logChecker,
  mapBetweenChecker
};